bin/generated_template.d.ts

```diff
@@ -0,0 +1,113 @@
+/** Provides information pointing to a location within a source. */
+export interface Location {
+  /** Line in the parsed source (1-based). */
+  line: number;
+  /** Column in the parsed source (1-based). */
+  column: number;
+  /** Offset in the parsed source (0-based). */
+  offset: number;
+}
+
+/** The `start` and `end` position's of an object within the source. */
+export interface LocationRange {
+  /** Any object that was supplied to the `parse()` call as the `grammarSource` option. */
+  source: any;
+  /** Position at the beginning of the expression. */
+  start: Location;
+  /** Position after the end of the expression. */
+  end: Location;
+}
+
+export interface LiteralExpectation {
+  type: "literal";
+  text: string;
+  ignoreCase: boolean;
+}
+export interface ClassParts extends Array<string | ClassParts> {
+}
+export interface ClassExpectation {
+  type: "class";
+  parts: ClassParts;
+  inverted: boolean;
+  ignoreCase: boolean;
+}
+export interface AnyExpectation {
+  type: "any";
+}
+export interface EndExpectation {
+  type: "end";
+}
+export interface OtherExpectation {
+  type: "other";
+  description: string;
+}
+export type Expectation =
+  | AnyExpectation
+  | ClassExpectation
+  | EndExpectation
+  | LiteralExpectation
+  | OtherExpectation;
+
+declare class _PeggySyntaxError extends Error {
+  /**
+   * Constructs the human-readable message from the machine representation.
+   *
+   * @param expected Array of expected items, generated by the parser
+   * @param found Any text that will appear as found in the input instead of
+   *   expected
+   */
+  static buildMessage(expected: Expectation[], found: string | null): string;
+  message: string;
+  expected: Expectation[];
+  found: string | null;
+  location: LocationRange;
+  name: string;
+  constructor(
+    message: string,
+    expected: Expectation[],
+    found: string | null,
+    location: LocationRange,
+  );
+  format(sources: {
+    source?: any;
+    text: string;
+  }[]): string;
+}
+
+export interface ParserTracer {
+  trace(event: ParserTracerEvent): void;
+}
+
+export type ParserTracerEvent
+  = { type: "rule.enter"; rule: string; location: LocationRange }
+  | { type: "rule.fail"; rule: string; location: LocationRange }
+  | { type: "rule.match"; rule: string; result: any; location: LocationRange };
```

Comment:
I think you might have been looking in the wrong place.  Generate a grammar with `--trace` turned on to see the code that will run.  Look for `peg$tracer.trace` calls.  The result is anything that a rule could return from an action block.  Example:

```peg.js
NaN = "nan" { return NaN }
```

Reply:
Aha! Yeah you're right, I see it in source code now too. My folly was checking for the string `"rule.match"` which only included the things I mentioned.

Looks like `unknown` is the best type for the job unless action blocks can't return, say, functions or something.

Reply:
I can totally imagine them returning a function.  For instance, you might implement an XPath parser that way...  (/me checks.  No, I didn't do that in my XPath parser, but it would have been reasonable)

Reply:
ClassParts

Reply:
I can't mark this as resolved but it can be.